### PR TITLE
feat(github)!: support integration with GitHub repos

### LINKED ISSUE #119: Add author/contributor name to the changelogs
**Is your feature request related to a problem? Please describe.**
Currently git-cliff does not parse out the author names and save them at the end of the changelog.

**Describe the solution you'd like**
Add an option to write out contributor names at the end of the changelog.

### PR DESCRIPTION
**Pull request by orhun**
Resolves #119 #132 #112 #12

## Description

This PR adds GitHub integration to `git-cliff`. The basic goal is to support generating the same changelog as the built-in GitHub changelog generator (with PR links and contributors). This means that we need to query the GitHub API for certain information and incorporate them in the changelog context.

## Usage

There is a new section in the config file for setting the GitHub remote:

```toml
[remote.github]
owner = "orhun"
repo = "git-cliff"
token = ""
```

These values are also passed to the template (via `additional_context` param of `Changelog::render` function):

```
**Full Changelog**: https://github.com/{{ remote.github.owner }}/{{ remote.github.repo }}/compare/{{ previous.version }}...{{ version }}\
```

There are new values in the changelog context which you can use to add GitHub usernames / pull request numbers to the changelog:

```
{{ commit.github.username }}
{{ commit.github.pr_number }}
```

Also, there is a new variable for contributors (`github.contributors`):

```
{% for contributor in github.contributors | filter(attribute="is_first_time", value=true) %}
  * @{{ contributor.username }} made their first contribution in #{{ contributor.pr_number }}
{%- endfor -%}
```

One cool thing, no network requests are made if no GitHub-related variables are used in the template.

## Example

`cargo run -- --unreleased --config examples/github.toml`

```md
## What's Changed

- feat(github): support integration with GitHub repos
- fix(changelog): set the correct previous tag when a custom tag is given by @orhun
- docs(website): add installation instructions for Homebrew by @sh-cho in #357
- feat(args): set `CHANGELOG.md` as default missing value for output option by @sh-cho in #354
- refactor(config): remove unnecessary newline from configs by @orhun

## New Contributors

- @sh-cho made their first contribution in #357

<!-- generated by git-cliff -->
```

## Implementation

There seems to be 2 ways of interacting with GitHub programmatically:

- REST API: works without a token but there is a rate limit. Has pagination.
- GraphQL: only works with a repository token. Has pagination.

I'm using the REST API for the current implementation and I'm fetching _all_ the commits and pull requests. This sounds suboptimal but unfortunately we need all the commits to generate the required release context. For example:

`git cliff -u`: only generates the changelog for the unreleased changes. If we want to have a section there as "first contributors" then we need the whole commit/PR history to check if the user has contributed before. Unfortunately GitHub contributors API does not give us the SHA1 of the commits that belong to the contributors so I'm using the pull requests API.

On top of that, there is pagination (max 100 entries in a page) so we need to send multiple requests to fetch all the history.

TL;DR: This is bad and it vastly slows down the changelog generation. But I don't know how to improve it. I have experimented with the GraphQL API but it also has pagination although you can filter stuff before hand. (also it requires a token at all times as mentioned before)

Ping top contributors @MarcoIeni @kenji-miyake @joshka @alerque for opinions/discussion.

## TODO List

- [x] Find out upstream remote from the Git repository (no configuration)
- [x] Add progress bar for network requests
- [x] Add caching to network requests (?)
- [x] Gate the implementation behind `github` feature (?)
- [x] Add new examples
  - [x] `github.toml`: same format as GitHub
  - [x] Grouped GitHub changelog, something like [Pixi's changelog](https://github.com/prefix-dev/pixi/blob/main/CHANGELOG.md)
- [x] Add command-line arguments for configuring the remote
  - [x] `--github-repo`
  - [x] `--github-token`
- [x] Add tests (unit/integration/fixtures)
- [x] Update documentation

### CODE REVIEW

**Comment by alerque** [comment]
I like the idea of this being available. I've wanted contributor and PR callouts in changelogs before and not having them has kept me manually writing changelogs for some projects.

Have you looked into testing whether `gh` can be used for some of the queries, or at least help with Auth issues when available to be able to use the GraphQL API? I just wonder how hard it would be to use those credentials *if found* and fall back to the Rest API if not. Besides that thought some sort of caching for the network results sounds like it might be in order, perhaps using Git's object storage?

If the features are not used if not configured my thought would be that it doesn't need gating, but then again for use in CI sometimes it's nice to be able to compile as trimmed down a binary as possible for the specific job at hand.

**Review by kenji-miyake — COMMENTED** [review]
(I noticed my comments added a few days ago are still pending. I'm submitting them now.)

Great feature! I added several comments that I found with a quick look.

File: git-cliff-core/src/release.rs [hunk 1]
```diff
@@ -25,9 +32,60 @@ pub struct Release<'a> {
...
+		// retain the commits that are not a part of this release for later on
+		// checking the first contributors.
+		github_commits.retain(|v| {
+			if let Some(commit) =
+				self.commits.iter_mut().find(|commit| commit.id == v.sha)
+			{
+				commit.github.username = Some(v.author.login.to_string());
+				commit.github.pr_number = github_pull_requests
+					.iter()
+					.find(|pr| pr.merge_commit_sha == Some(v.sha.clone()))
```
[the remark this answers is not in the document]
Comment (orhun) [reply git-cliff-core/src/release.rs:63]: For me it is like this:

```
## What's Changed
* docs(lib): update the description of the error module
* feat(github): support deriving upstream URL from the repo
* fix(test): update tests accordingly to the render parameters
* refactor(github): use verbose logging for errors
* fix(changelog): fix generation logic
* Merge branch 'main' into feat/github_integration
* feat(parser): support using regex scope values by @orhun in #372
* test(fixture): update the date for example test fixture by @orhun
* docs(fixtures): add instructions for adding new fixtures by @orhun
* feat(args): support initialization with built-in templates by @orhun in #370
* feat(changelog)!: support templating in the footer by @orhun in #369
* feat(args): allow returning the bumped version by @nappa85 in #362
* test(fixture): add test fixture for bumping version by @orhun in #360
* fix: allow version bump with a single previous release by @nappa85
* feat(github): support integration with GitHub repos
* fix(changelog): set the correct previous tag when a custom tag is given by @orhun
* docs(website): add installation instructions for Homebrew by @sh-cho in #357
* feat(args): set `CHANGELOG.md` as default missing value for output option by @sh-cho in #354
* refactor(config): remove unnecessary newline from configs by @orhun

## New Contributors
* @nappa85 made their first contribution in #362
* @sh-cho made their first contribution in #357

<!-- generated by git-cliff -->
```

Can you try with `-vv` and see if there is anything useful to debug this?

[the remark this answers is not in the document]
Comment (orhun) [reply git-cliff-core/src/release.rs:63]: Hmm, maybe it is because that user has 2 PRs and I'm using `HashSet` which might not take the 2nd PR into account 🤔

[the remark this answers is not in the document]
Comment (orhun) [reply git-cliff-core/src/release.rs:63]: Ok I found out why. PR 360 was not squashed and it was rebased so `update_github_metadata` fails to find the PR number associated with the commit due to this logic:

```rust
find(|pr| pr.merge_commit_sha == Some(v.sha.clone()))
```

Instead, it should be something like this:

```rust
find(|pr| pr.commits.contains(v.sha.clone()))
```

Which means I need to send additional requests to `repos/<owner>/<repo>/pulls/<pr_number>/commits` to fetch all the commits in the PR.

🤮

**Comment by alerque** [comment]
> Just tried that and `gh` also has pagination which means it does not really help with anything. Fallback to `gh` means additional command execution so I would rather do that directly with Rust. Good idea though, thanks for sharing it.

Makes sense to not add external dependencies if it doesn't actually gain you anything functionally or performance wise.

**Comment by joshka** [comment]
bad bot :D

**Comment by alerque** [comment]
Yes. This bot has to be the new second most annoying thing hitting my notifications (right behind stale bots). I doesn't mean anything interesting at all happened to a project I'm following but besides the little user icon there is no way to know that from the notifications page.

**Comment by orhun** [comment]
Sorry for the notifications guys - pls don't unsub, will fix the bot.

**Comment by alerque** [comment]
It's too bad GH notification control isn't a bit more granular. "Ignore dependabot PRs for projects I'm not a collaborator on" or "never notify anybody for mergify pushes in my projects" are options I'd love to see.

File: git-cliff-core/src/release.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (orhun) [reply git-cliff-core/src/release.rs]: Well no, I can simply use the GitHub username for linking a PR which is what I exactly did in e44ebbc73008112c2434ca79d53ed9823df62a58

So there is no need to use `repos/<owner>/<repo>` API.

@kenji-miyake can you try running the same command again?

[the remark this answers is not in the document]
Comment (orhun) [reply git-cliff-core/src/release.rs]: Wait no, that broke eveything 💀 I will find another workaround.

**Comment by orhun** [comment]
Added a progress bar, but it is still slow imo:

rec_20231216T184225

**Comment by alerque** [comment]
The progress bar is a nice touch for something that takes so long. And yes it is a bit on the slow side but for integrating external data and being at the mercy of a remote paged API it is somewhat expected/bearable. Any contributions to make it faster would be appreciated, but I wouldn't consider the lack-luster speed a blocker at this point.

File: git-cliff-core/src/github.rs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+			HeaderValue::from_static("application/vnd.github+json"),
+		);
+		if let Some(token) = remote.token {
+			headers.insert(
+				reqwest::header::AUTHORIZATION,
+				format!("Bearer {}", token.expose_secret()).parse()?,
+			);
+		}
+		headers.insert(reqwest::header::USER_AGENT, USER_AGENT.parse()?);
+		let client = Client::builder()
```
Comment (tomverelst) [inline git-cliff-core/src/github.rs:176]: It is possible to keep the TCP connection open using https://docs.rs/reqwest/latest/reqwest/struct.ClientBuilder.html#method.tcp_keepalive, so that might 

Normally it is 60 seconds by default, but it is [disabled](https://docs.rs/reqwest/latest/src/reqwest/async_impl/client.rs.html#186) at the moment.

**Comment by orhun** [comment]
@alerque yeah, I totally agree. With the HTTP cache + client optimizations, it is much faster for small repos (800 commits / 300 PRs):

rec_20231218T161759

**Comment by orhun** [comment]
I added the documentation: https://github.com/orhun/git-cliff/blob/feat/github_integration/website/docs/integration/github.md

**Comment by alerque** [comment]
Off topic of this PR but on the topic of conventional commits and generated changelogs: my understanding of `fix:` is that it is only relevant if the commit fixes something in the previous release version, not if it fixes some feature addition that hasn't been released yet. As an example I would have used `chore: ` for 32858b7 so that when I generate a change log for the next release it does not show up (I also typically don't even list chore commits in the changelog). This is almost certainly a personal preference to have a whole class of commits not listed at all, but not having things in the "bug fixes" section that are not bugs in the previous release(s) seems to make sense to me. Is your usage in this PR deliberate to include fixes of the in-progress feature or no?

**Comment by orhun** [comment]
> Is your usage in this PR deliberate to include fixes of the in-progress feature or no?

Yes, the changes that are in this PR is only scoped this feature. I think I'm going to squash the commits with a message such as "feat!: add github integration" or something.

**Comment by orhun** [comment]
alright, here we go
